view/log.go

```diff
@@ -149,6 +149,136 @@ func (li *logView) Read(facts origins.Facts) (int, error) {
 	return l, nil
 }
 
+// merger maintains state of a merged stream of facts from multiple domains.
+type merger struct {
+	iterators []origins.Iterator
+	next      []*origins.Fact
+}
+
+func (m *merger) Next() (*origins.Fact, error) {
+	var maxTrn uint64 = 0
+	var maxI int = -1
+
+	// look at the next fact coming out of each of the the available logs,
+	// and choose the fact with the greatest transaction ID.
+	for i := 0; i < len(m.iterators); i++ {
+		if m.next[i] != nil && m.next[i].Transaction > maxTrn {
+			maxTrn = m.next[i].Transaction
+			maxI = i
+		}
+	}
+
+	if maxI == -1 {
+		return nil, io.EOF
+	}
+
+	// push the newest fact out of the merged stream
+	nextFact := m.next[maxI]
+	fact, err := m.iterators[maxI].Next()
+
+	if err != nil {
+		if err == io.EOF {
+			m.next[maxI] = nil
+		} else {
+			return nil, err
+		}
+	} else {
+		m.next[maxI] = fact
+	}
+
+	return nextFact, nil
+}
+
+func (m *merger) init() error {
+	var err error
+
+	for i := 0; i < len(m.iterators); i++ {
+		m.next[i], err = m.iterators[i].Next()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// merge multiple fact streams into one.
+// Note: iterators return facts in the order opposite to that in which
+// they were committed (i.e. newest facts with largest transaction IDs first),
+// and the Merge operation will preserve that.
+func Merge(iterators []origins.Iterator) (*merger, error) {
+	m := merger{
+		iterators: iterators,
+		next:      make([]*origins.Fact, len(iterators)),
+	}
+
+	err := m.init()
+	if err != nil {
+		return nil, err
+	}
+
+	return &m, nil
+}
+
+// deduplicator maintains the state of a steam of unique facts
+type deduplicator struct {
+	iter          origins.Iterator
+	observedFacts *FactMap
+}
+
+func (d *deduplicator) Next() (*origins.Fact, error) {
+	for {
+		f, err := d.iter.Next()
+
+		if err != nil {
+			return nil, err
+		}
+
+		if !d.observedFacts.add(f) {
+			return f, nil
+		}
+	}
+}
+
+// remove duplicate facts from a stream
+func Deduplicate(iter origins.Iterator) *deduplicator {
```

Comment:
Same as above, return a `origins.Iterator` and improve the documentation for this function.